### PR TITLE
Allow checks to be performed inside quantifier statements.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This allows checks from CLI flags like `--pointer-check` to be performed inside
expressions in quantifier statements.

This should be addressing issue #6231 .

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — COMMENTED** [review]
I can't immediately see anything wrong with this BUT the whole thing makes me uneasy.  The comments have been there since 2013 which makes me think that they were probably there for a reason.  My immediate thought is about scoping and whether the quantified variable is actually live when the variable is checked.  I am also unsure about how nested quantifiers are going to work and to what degree the guards are applied to the checks as well.

Can you please say a little more about how you are expecting things to work?

File: regression/cbmc-primitives/forall_6231_2/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_2/test.c:14]: Likewise; not just one instantiation please.

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_1/test.c:16]: Please can you change this so that it doesn't just have one true instantiation?

File: regression/cbmc-primitives/forall_6231_2/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_2/test.c:20]: Again, multiple instantiations.

File: regression/cbmc-primitives/forall_6231_2/test.desc
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_2/test.desc:20]: Please can we have one test that fails?

**Review by martin-cs — CHANGES_REQUESTED** [review]
A few more concrete requests for other functionality.

**Review by kroening — CHANGES_REQUESTED** [review]
This happens to work for the specific case of positive forall, since this is doing Skolemisation implicitly.

1. This will fail (generating VCCs that are too strong) for exists.
2. I would much prefer that the Skolemisation is done explicitly, i.e., by introducing a Skolem constant and instantiating the quantifier with it.

File: regression/cbmc-primitives/forall_6231_2/test.c
Comment (NlightNFotis) [reply regression/cbmc-primitives/forall_6231_2/test.c:14]: Hi @martin-cs, I've left these as they stand, and added a separate test that does multiple instantiations (by essence of it having a bigger upper bound).

Does that sound like an adequate solution to you?

File: regression/cbmc-primitives/forall_6231_2/test.desc
Comment (NlightNFotis) [reply regression/cbmc-primitives/forall_6231_2/test.desc:20]: By a similar token, I've added a test that has the `pointer being outside of object bounds` condition fail by having the `malloc_size < upper_bound`.

Is that adequate, or should I spend some more time thinking about it?

Comment (martin-cs) [reply regression/cbmc-primitives/forall_6231_2/test.desc:20]: Can I see the extra tests before I comment on them?

Comment (NlightNFotis) [reply regression/cbmc-primitives/forall_6231_2/test.desc:20]: Yes, of course - I've added them as a separate commit (the third one).

**Comment by NlightNFotis** [comment]
Hi Daniel @kroening.

With regards to your first query, we have pushed some changes that show that existentially quantified statements appear to be working as they should (or at least, how we expect them to). Could you review these and let us know if these much your expectations, or if there's a counterexample case you could think of that would help us test our change in a way that exposes a weakness of our current implementation?

File: regression/cbmc-primitives/exists_6231_2/test.c
Comment (TGWDB) [inline regression/cbmc-primitives/exists_6231_2/test.c]: This comment does not look correct.

File: regression/cbmc-primitives/exists_6231_2/test.desc
Comment (TGWDB) [inline regression/cbmc-primitives/exists_6231_2/test.desc]: Does this check the quantified expression, or will it "pass" when the assert on line 9 fails?

File: regression/cbmc-primitives/forall_6231_3/test.c
Comment (TGWDB) [inline regression/cbmc-primitives/forall_6231_3/test.c]: This comment is false?

File: regression/cbmc-primitives/forall_6231_3/test_malloc_less_than_bound.c
Comment (TGWDB) [inline regression/cbmc-primitives/forall_6231_3/test_malloc_less_than_bound.c]: False comment?

**Review by TGWDB — CHANGES_REQUESTED** [review]

File: regression/cbmc-primitives/exists_6231_2/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/exists_6231_2/test.c]: Why lots of errors; the only issue I can see is that `*a` is uninitialised.

File: regression/cbmc-primitives/exists_6231_2/test.desc
Comment (martin-cs) [reply regression/cbmc-primitives/exists_6231_2/test.desc]: Yeah; this is too general.  Is there a reason why you are no longer checking each of the VCs gives the right answer?

File: regression/cbmc-primitives/exists_memory_checks/invalid_index_range.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/invalid_index_range.desc]: OK; this is encouraging.

File: regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc]: Where does the `0 <= i && i < 20` come from?

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_1/test.c]: Again, why?

File: regression/cbmc-primitives/forall_6231_2/test.c
Comment (martin-cs) [reply regression/cbmc-primitives/forall_6231_2/test.c:14]: It is OK but I think it might make these tests redundant.

File: regression/cbmc-primitives/forall_6231_2/test.desc
Comment (martin-cs) [reply regression/cbmc-primitives/forall_6231_2/test.desc:20]: Thanks!

**Review by martin-cs — CHANGES_REQUESTED** [review]
1. Do you have examples of quantifiers in assumptions?
2. Can you briefly outline how this works?  What logical expression is generated?  How does this interact with scoping?

File: regression/cbmc-primitives/exists_6231_2/test.c
Comment (SaswatPadhi) [reply regression/cbmc-primitives/exists_6231_2/test.c]: In [my original example](#6231), `a` was uninitialized so `*a` was resulting in lots of issues (with `--pointer-check`).

I think we could remove this comment now that we `malloc` above.

File: regression/cbmc-primitives/forall_6231_3/test.c
Comment (SaswatPadhi) [reply regression/cbmc-primitives/forall_6231_3/test.c]: In [my original example](#6231), `a` was uninitialized so `*a` was resulting in lots of issues (with `--pointer-check`).

I think we could remove this comment now that we `malloc` above.

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (SaswatPadhi) [reply regression/cbmc-primitives/forall_6231_1/test.c]: In [my original example](#6231), `a` was uninitialized so `*a` was resulting in lots of issues (with `--pointer-check`).

I think we could remove this comment now that we `malloc` above.

**Review by SaswatPadhi — COMMENTED** [review]
Thanks for fixing this! This would help us catch issues in code contracts. Earlier, invariants and function contracts that used implications but performed incorrect/unsafe operations were not caught.

All test cases in this PR seem to check that the consequent is safe. May be the fix already checks the antecedent as well, but can we also add a regression test for that? Something like:
```c
__CPROVER_forall {
  int i;
  (0 <= i && i <= (n / 0))    /* (n/0) should be caught by `--div-by-zero` */
  ==> ...
}
```

**Comment by martin-cs** [comment]
@SaswatPadhi makes a good point.  It is not clear to me whether this treats quantified expressions as guard/statement pairs and if so, whether it checks in both, does it use the guard for the checks in the statement?

File: regression/cbmc-primitives/exists_6231_2/test.c
Comment (thomasspriggs) [reply regression/cbmc-primitives/exists_6231_2/test.c]: This test was from a work in progress commit, which should have now been dropped.

File: regression/cbmc-primitives/exists_6231_2/test.desc
Comment (thomasspriggs) [reply regression/cbmc-primitives/exists_6231_2/test.desc]: This test was from a work in progress commit, which should have now been dropped.

File: regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc
Comment (NlightNFotis) [reply regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc]: This was a mistake - the test was checking against the wrong test file.

**Comment by NlightNFotis** [comment]
> @SaswatPadhi makes a good point. It is not clear to me whether this treats quantified expressions as guard/statement pairs and if so, whether it checks in both, does it use the guard for the checks in the statement?

Hi Martin, yes, it does treat quantified expressions as guard/statement pairs and yes it uses the guard for the checks in the statement. I have factored the change that does that in [its own commit](https://github.com/diffblue/cbmc/pull/6399/commits/1cdc6e5fed1ba60635d10f5b175612a406f32d0e), with an explanation of how it works there.

As a shorter explanation here, `goto_checkt::check` in the same file calls `check_rec` on the expression with `true_exprt` as a guard, which is then refined based on the exact nature of the expression. We already have support for refining the guard for logical operators, but we were dispatching to that only if the expression had `expr.id() == ID_and` or `expr.id() == ID_or`. This appears to be a latent bug that was not manifest because we didn't have any expressions of the particular form that the tests submitted for this PR, so we didn't observe that. As soon as we started performing checks for quantifier expressions (basically, not immediately bailing upon seeing one), we hit this bug.

Does that answer your question?

**Comment by NlightNFotis** [comment]
Hi @kroening ,

I'm trying to better understand the requirement of the second part of your comment, which I quote:

> I would much prefer that the Skolemisation is done explicitly, i.e., by introducing a Skolem constant and instantiating the quantifier with it.

My understanding is that Skolemization is being done the same way it was being done before this patch, i.e. in `goto_symext::symex_assert()`  of `src/goto-symex/symex_main.cpp:157`, which appears to convert existential quantifiers into universals (the first part of the Skolemization process) and then performs a renaming of the bound variable (which performs the substitution by a Skolem function/Skolem constant).

Is that not adequate in some capacity? We have performed extended testing and it seems like this bug fix is working as intended - but if you believe that there's a chance some input might fail we're happy to investigate further and tighten the coverage of the tests for this.

Or is your request that we lift this all into a separate module? If yes, where do you think this would be better plugged in - the simplification modules, a completely different pass over the goto-program that hooks into the transformation pipeline, or somewhere else entirely?

**Review by TGWDB — APPROVED** [review]

**Comment by martin-cs** [comment]
@NlightNFotis Thanks for your response, the explanation does indeed help and yes, I think you are correct about this being a latent bug.  Would it be possible to write a test case that doesn't use quantifiers that triggers this?  Would something like `y != 0 ==> 10/y <= 10` and `--div-by-zero-check` do?  That way the fixes could be split from the new feature and be approved even if there is more discussion over the checks in quantifiers.

Please could you have a look at, check and if necessary update:
https://github.com/diffblue/cbmc/blob/ce2680b724d650c404468eb0070c8b402735ec66/src/analyses/goto_check.cpp#L111
https://github.com/diffblue/cbmc/blob/ce2680b724d650c404468eb0070c8b402735ec66/src/analyses/goto_check.cpp#L1623

( The real cherry on top would be adding a `PRECONDITION` that checks for the expression types that are intended to be supported. )

**Comment by kroening** [comment]
The problem is that the identifiers used in quantifiers are no longer guaranteed to be unique, i.e., you could have something like
`i[2]== 2 && exists i:int. x/i`
which will yield a check that reads
`(i[2] != 2) => i!=0`
In this formula, `i` is used twice, and isn't even type consistent -- the first instance is an array, and the second is an int. Formulas like that must not be generated.

You can't test that in the C front-end, since the C front-end's treatment of quantifiers pre-dates the introduction of scoping in the solver backends. Please don't even think about reintroducing this restriction, not even as a temporary hack.

The way to do this properly is to introduce a Skolem symbol for the bound identifier. You'll find that the classes for exists and forall have convenient methods for making that easy!

**Comment by martin-cs** [comment]
@NlightNFotis while reviewing the changes you had kindly described I ran into some concerns.  I have documented them here: https://github.com/diffblue/cbmc/issues/6423  I don't consider them a precondition for this PR, but, obviously, they are linked.

**Comment by NlightNFotis** [comment]
@kroening Hi Daniel, the [last commit](https://github.com/diffblue/cbmc/pull/6399/commits/63f683069c1e2096923e732329e74edf33aba048) is performing scolemisation of the formula in `goto_check`. Can you please review it now and let us know if this is adequate for what you had in mind?

@martin-cs Given that the bug-fix has now been merged on its own PR, could you please re-review this and let me know if there's anything more that needs be done?

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: Clang-tidy suggests this should be `const auto &` to avoid unnecessary copy construction.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: Given that this code is not part of the C front end, hard coding the language to C seems like the wrong choice. I suggest using the language mode of `quantifier_sym`.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: Don't cast away const! `get_fresh_aux_symbol` mutates the symbol table by adding symbols. Therefore casting away const violates the promise `goto_checkt` makes to not mutate the symbol table / by declaring them as const.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

**Comment by NlightNFotis** [comment]
Hi @kroening, can you take a look at the PR now and let us know if it is more or less what you envisioned for skolemization as a pass?

Specifically, it's the last commit that implements this.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [inline src/goto-programs/skolemization.cpp]: :pick: `recurse_on_` just tells the reader how this function is implemented, not what it does with/to conditions.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [reply src/analyses/goto_check.cpp]: No longer relevant.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [inline src/goto-programs/skolemization.cpp]: :no_entry: I find this comment really misleading. The following statement will be reached whether or not this is a leaf node. In fact the nodes of interest are not leaf nodes, as they are exists quantifiers and these quantifier expressions have child nodes. I think the phrase you are after is "depth first search".

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [reply src/analyses/goto_check.cpp]: This is still hard coded in your new `rename_and_substitute` function and still needs fixing.

File: src/goto-programs/skolemization.h
Comment (thomasspriggs) [inline src/goto-programs/skolemization.h]: :pick: As this `recurse_on_condition` function and the `rename_and_substitute` function below are implementation details, I would not expose them as part of the interface in the header file. Note that functions can be forward declared in the `.cpp` file or re-ordered to avoid needing to forward declare at all.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [inline src/goto-programs/skolemization.cpp]: Using recursion like this to visit all the nodes in the condition is re-implementing already standardised visitation. This should really be done with `exprt::visit_post`. I have tried out this refactor here - https://github.com/NlightNFotis/cbmc/compare/fix_6231...thomasspriggs:tas/fix_6231

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [inline src/goto-programs/skolemization.cpp]: :bulb: Just use `symbolt::symbol_expr()`. Its a lot shorter and easier to read for me.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [inline src/goto-programs/skolemization.cpp]: :pick: Includes of `<ansi-c/expr2c.h>` and `<util/expr_util.h>` appear to be unused.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [reply src/analyses/goto_check.cpp]: This looks fixed.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [reply src/goto-programs/skolemization.cpp]: This comment is now gone.

File: src/goto-programs/skolemization.h
Comment (thomasspriggs) [reply src/goto-programs/skolemization.h]: Ok.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [reply src/goto-programs/skolemization.cpp]: Ok.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [reply src/goto-programs/skolemization.cpp]: Ok, this is now resolved.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [reply src/goto-programs/skolemization.cpp]: Ok, this is now resolved.

File: src/goto-programs/skolemization.cpp
Comment (thomasspriggs) [reply src/goto-programs/skolemization.cpp]: Ok, this is now resolved.

**Review by thomasspriggs — APPROVED** [review]

**Comment by NlightNFotis** [comment]
Hello @martin-cs and @kroening , would it be possible to have a look at this again?

It's past our internal reviews and blocked on your reviews right now. I want to take this forward, either in terms of performing improvements and revisions if deemed necessary, or getting this merged.

**Comment by martin-cs** [comment]
@NlightNFotis My apologies for the delay in reviewing.  I will look ASAP.

File: regression/cbmc-primitives/exists_memory_checks/invalid_index_range.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/invalid_index_range.desc:8]: Would it be possible to check that the assertion passes?

File: regression/cbmc-primitives/forall_6231_2/test.c
Comment (martin-cs) [reply regression/cbmc-primitives/forall_6231_2/test.c:20]: Ditto.

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (martin-cs) [reply regression/cbmc-primitives/forall_6231_1/test.c:16]: I see you have added this as a separate test.  OK.

File: regression/cbmc-primitives/exists_memory_checks/negated_exists.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/negated_exists.desc]: A more complete check line as you have in other tests would be good.

File: regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc:3]: As we still don't have a CI run that doesn't have Z3 set up, please can this be tagged so that the regression tests still work on machines that don't have Z3 installed?

File: regression/cbmc-primitives/exists_memory_checks/valid_index_range.desc
Comment (martin-cs) [inline regression/cbmc-primitives/exists_memory_checks/valid_index_range.desc]: More complete matching against results would be good.

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_1/test.c]: I am not quite sure what this comment is trying to tell me.  Whatever it is, it feels like there might be a better place and way or saying it.

File: regression/cbmc-primitives/forall_6231_3/test_malloc_less_than_bound.c
Comment (martin-cs) [inline regression/cbmc-primitives/forall_6231_3/test_malloc_less_than_bound.c:14]: As above.

File: src/goto-programs/skolemization.cpp
Comment (martin-cs) [inline src/goto-programs/skolemization.cpp]: Thank you!

File: src/goto-programs/skolemization.cpp
Comment (martin-cs) [inline src/goto-programs/skolemization.cpp]: This is not what I would think of as Skolemization; this is renaming.  For me, Skolemization has to remove the existential so:

```
assert(__CPROVER_exists {  int i;   (0 <= i && i < 10) && a[i] == i *i   });
```

becomes:

```
int skolem_23_i;  // Note that if it is nested in a forall then it will need to be an uninterpreted function.
assert( (0 <= skolem_23_i && skolem_23_i < 10) && a[skolem_23_i] == skolem_23_i * skolem_23_i );
```

The general use-case for this is so that provers only have to deal with one kind of quantifier.

Note that the symex engine performs on-the-fly Skolemization where it can; that's how the SAT engine can handle quantifiers at all.  However it dynamically handles polarity, so depending on whether the quantifiers are in assume or assert, negated, etc. it will do the right thing.

**Review by martin-cs — COMMENTED** [review]
I am removing my block on this but I still have some concerns:

* There are no tests using `assume`; that matters from a user point of view because of code contracts and also from a technical point of view as these are "flipped" during symex so what works with the SAT engine is different.
* A lot of the issues with quantifiers are simplified when there is only one level of quantifier alternation.  So I feel that alternating tests would be useful.
* The renaming code is not, IMHO, Skolemization and to me, doesn't match what @kroening requested.  However I believe that it fixes the immediate issue he raised and I will leave it to him to pursue that one.
* I'm still struggling a bit to see what is generated; I /think/ it is something like

`A && __CPROVER_forall { int i; B ==> C} `

will create a VC:

`assert(A && __CPROVER_forall { int i; B ==> CHECK_FROM_C } ); `

it would be good to get that confirmed.

**Review by martin-cs — APPROVED** [review]
Annoying -- github won't let me just remove my "Request changes" and return to a neutral state.  So, I am "approving" this but please don't treat this approved and merge.

**Comment by kroening** [comment]
Please look at https://github.com/diffblue/cbmc/tree/goto_check_guard for an alternative approach, which relegates the problem of introducing the Skolem functions to the solver.

A key benefit of the alternative approach is that this means that there won't be two different kinds of goto-programs, one where bound symbols are unique and one where they aren't.

File: src/goto-programs/skolemization.cpp
Comment (NlightNFotis) [reply src/goto-programs/skolemization.cpp]: Hi Martin,

Apologies for that - I'm aware of how skolemization works (in contrast to the code here), but when implemented in its proper form, it resulted in elimination of the existential quantifier far too early in the pipeline. There's code in `goto-check` or `goto-symex` (can't remember which of the two) - which depended on the quantifiers being there during VCC generation, and we were left with vccs that didn't semantically match the intended meaning, getting results that were all over the place.

The intended effect of the code above was to do a partial-skolemization, if we can admit that as a thing: perform the renaming, and let the existential quantifier stay, with the result being that the correct VCCs where generated, and it was being eliminated later in the pipeline in a way that felt safer (given that the variables were renamed before hand).

Definitely wasn't ideal, but there was a lot of uncertainty around the implementation of that feature in our pipeline.

My aim now is to drop this and substitute for Daniel's implementation (assuming it passes all tests).

File: regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc
Comment (NlightNFotis) [reply regression/cbmc-primitives/exists_memory_checks/smt_missing_range_check.desc:3]: Yes, in the last commit I have added a check for the `z3` binary in `CMakeLists.txt`

File: regression/cbmc-primitives/forall_6231_1/test.c
Comment (NlightNFotis) [reply regression/cbmc-primitives/forall_6231_1/test.c]: Apologies, I added this comment in a haste, and now realise that my phrasing was not the best 😞 

I have rephrased that in commit 604d12c. Can you let me know if that's clearer?

**Comment by NlightNFotis** [comment]
Hi @martin-cs and @kroening , 

as requested, I've made several enhancements to the tests (added the extra checks Martin requested, tests based on the behaviour of the quantifiers in assumes, and alternating quantifier tests - each of which are in their own commits) and changed the implementation to be using Daniel's code (I lost the commit information because I couldn't cleanly cherry-pick after some changes to `goto-check`, but I hope my commit message preserves authorship information in an acceptable way).

Could you please have a look again and let me know how it looks to you now?

**Comment by kroening** [comment]
May I suggest squashing the 9 commits that add/change the tests? Otherwise looks fine.

**Comment by NlightNFotis** [comment]
Hi @kroening, if all is fine are you happy to lift your blocking review?

I will of course squash all the test commits before the final merge, I just want to give a chance to @martin-cs to review them on a commit by commit basis so that he can more easily check if his concerns have been addressed accordingly, and assuming he's also happy I will squash and merge.

**Review by kroening — APPROVED** [review]

**Pull request merged**
